File: python/src/attesta/exporters.py
"""Audit trail exporters for CSV and JSON formats.

Provides :class:`CSVExporter` and :class:`JSONExporter` that implement the
:class:`AuditExporter` protocol for exporting audit entries to standard
formats.

Usage::

    from attesta.core.audit import AuditLogger
    from attesta.exporters import CSVExporter

    audit = AuditLogger(path=".attesta/audit.jsonl")
    entries = audit.query(verdict="approved")

    with open("report.csv", "w") as f:
        CSVExporter().export(entries, f)
"""

from __future__ import annotations

import csv
import json
from typing import IO, Any, Protocol, runtime_checkable

from attesta.core.audit import AuditEntry

__all__ = ["AuditExporter", "CSVExporter", "JSONExporter"]

# Default columns for CSV export.
DEFAULT_COLUMNS: list[str] = [
    "entry_id",
    "intercepted_at",
    "action_name",
    "risk_score",
    "risk_level",
    "challenge_type",
    "verdict",
    "agent_id",
    "review_duration_seconds",
    "chain_hash",
]


@runtime_checkable
class AuditExporter(Protocol):
    """Protocol for audit trail exporters."""

    def export(self, entries: list[AuditEntry], output: IO[str]) -> None:
        """Write *entries* to *output* in the exporter's format."""
        ...


class CSVExporter:
    """Export audit entries as CSV.

    Parameters
    ----------
    columns:
        List of field names to include as columns. Defaults to
        :data:`DEFAULT_COLUMNS`. Nested dict fields (like ``metadata``)
        are JSON-serialised.
    """

    def __init__(self, columns: list[str] | None = None) -> None:
        self.columns = columns or list(DEFAULT_COLUMNS)

    def export(self, entries: list[AuditEntry], output: IO[str]) -> None:
        """Write *entries* to *output* as CSV."""
        writer = csv.writer(output)
        writer.writerow(self.columns)

        for entry in entries:
            row_dict = entry.to_dict()
            row: list[str] = []
            for col in self.columns:
                val = row_dict.get(col, "")
                # JSON-serialise complex types (dicts, lists)
                if isinstance(val, (dict, list)):
                    val = json.dumps(val, separators=(",", ":"))
                row.append(str(val) if val is not None else "")
            writer.writerow(row)


class JSONExporter:
    """Export audit entries as a JSON array.

    Parameters
    ----------
    indent:
        Number of spaces for pretty-printing. ``None`` for compact output.
    """

    def __init__(self, indent: int | None = 2) -> None:
        self.indent = indent

    def export(self, entries: list[AuditEntry], output: IO[str]) -> None:
        """Write *entries* to *output* as a JSON array."""
        data = [entry.to_dict() for entry in entries]
        json.dump(data, output, indent=self.indent, default=str)
        output.write("\n")

File: typescript/src/exporters.ts
/**
 * Audit trail exporters for CSV and JSON formats.
 *
 * @example
 * ```ts
 * import { AuditLogger } from "@kyberon/attesta";
 * import { CSVExporter } from "@kyberon/attesta/exporters";
 *
 * const audit = new AuditLogger(".attesta/audit.jsonl");
 * const entries = audit.query({ verdict: "approved" });
 * const csv = new CSVExporter().export(entries);
 * ```
 */

import type { AuditEntryData } from "./audit.js";

// ---------------------------------------------------------------------------
// AuditExporter interface
// ---------------------------------------------------------------------------

/**
 * Interface for audit trail exporters.
 */
export interface AuditExporter {
  /** Export entries and return as a string. */
  export(entries: AuditEntryData[]): string;
}

// ---------------------------------------------------------------------------
// Default columns
// ---------------------------------------------------------------------------

export const DEFAULT_COLUMNS: readonly string[] = [
  "entryId",
  "interceptedAt",
  "actionName",
  "riskScore",
  "riskLevel",
  "challengeType",
  "verdict",
  "agentId",
  "reviewDurationSeconds",
  "chainHash",
] as const;

// ---------------------------------------------------------------------------
// CSVExporter
// ---------------------------------------------------------------------------

/**
 * Export audit entries as CSV.
 */
export class CSVExporter implements AuditExporter {
  readonly columns: readonly string[];

  constructor(options?: { columns?: string[] }) {
    this.columns = options?.columns ?? DEFAULT_COLUMNS;
  }

  export(entries: AuditEntryData[]): string {
    const lines: string[] = [];

    // Header
    lines.push(this.columns.join(","));

    // Rows
    for (const entry of entries) {
      const record = entry as unknown as Record<string, unknown>;
      const row = this.columns.map((col) => {
        const val = record[col];
        if (val == null) return "";
        if (typeof val === "object") return csvEscape(JSON.stringify(val));
        return csvEscape(String(val));
      });
      lines.push(row.join(","));
    }

    return lines.join("\n") + "\n";
  }
}

// ---------------------------------------------------------------------------
// JSONExporter
// ---------------------------------------------------------------------------

/**
 * Export audit entries as a JSON array.
 */
export class JSONExporter implements AuditExporter {
  readonly indent: number | undefined;

  constructor(options?: { indent?: number }) {
    this.indent = options?.indent ?? 2;
  }

  export(entries: AuditEntryData[]): string {
    return JSON.stringify(entries, null, this.indent) + "\n";
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function csvEscape(value: string): string {
  if (
    value.includes(",") ||
    value.includes('"') ||
    value.includes("\n")
  ) {
    return `"${value.replace(/"/g, '""')}"`;
  }
  return value;
}
